util/dvsim/NcLauncher.py

```diff
@@ -0,0 +1,217 @@
+                    e, self.deploy.get_log_path()))
+            finally:
+                self._close_process()
+        else:
+            # Interactive: Set RUN_INTERACTIVE to 1
+            exports['RUN_INTERACTIVE'] = '1'
+
+            # Interactive. stdin / stdout are transparent
+            # no timeout and blocking op as user controls the flow
+            print('Interactive mode is not supported yet.')
+            print('Cmd : {}'.format(self.deploy.cmd))
+            self.process = subprocess.Popen(shlex.split(self.deploy.cmd),
+                                            stdin=None,
+                                            stdout=None,
+                                            stderr=subprocess.STDOUT,
+                                            # string mode
+                                            universal_newlines=True,
+                                            env=exports)
+
+            # Wait until the process exit
+            self.process.wait()
+
+        self._link_odir('D')
+
+    def poll(self):
+        """Check status of the running process.
+
+        This returns 'D', 'P', 'F', or 'K'. If 'D', the job is still running.
+        If 'P', the job finished successfully. If 'F', the job finished with
+        an error. If 'K' it was killed.
+
+        This function must only be called after running self.dispatch_cmd() and
+        must not be called again once it has returned 'P' or 'F'.
+        """
+
+        assert self.process is not None
+        elapsed_time = datetime.datetime.now() - self.start_time
+        self.job_runtime_secs = elapsed_time.total_seconds()
+        if self.process.poll() is None:
+            if self.timeout_secs and (self.job_runtime_secs >
+                                      self.timeout_secs) and not self.deploy.gui:
+                self._kill()
+                timeout_message = 'Job timed out after {} minutes'.format(
+                    self.deploy.get_timeout_mins())
+                self._post_finish(
+                    'K',
+                    ErrorMessage(line_number=None,
+                                 message=timeout_message,
+                                 context=[timeout_message]))
+                return 'K'
+
+            return 'D'
+
+        self.exit_code = self.process.returncode
+        status, err_msg = self._check_status()
+        self._post_finish(status, err_msg)
+        return self.status
+
+    def _kill(self):
+        """Kill the running process.
+
+        Try to kill the running process. Send SIGTERM first, wait a bit,
+        and then send SIGKILL if it didn't work.
+        """
+        assert self.process is not None
+        subprocess.call(['nc', 'stop', '-set', self.deploy.full_name, '-sig',
+                         'TERM,KILL'])
+        self.process.terminate()
+        try:
+            self.process.wait(timeout=2)
+        except subprocess.TimeoutExpired:
+            self.process.kill()
```

Comment:
Can `self.process` ever still be running after the `nc stop` command above has run to completion? I don't have experience with `nc` (and my knowledge of this is from vague memories of LSF!), but I'm slightly surprised that this is needed.

In either case, can you add a comment to the code that explains how the dance is supposed to work?

Reply:
I originally copied this section of code from sgelauncher and sge didn't have an explicit stop/kill command. after going through your comments, I referred to the LsfLauncher which is more closer to this implementation since both lsf and nc have stop/kill commands. I was able to get rid of the self.process.terminate and related code after that